src/dawn/IIR/DoMethod.cpp

```diff
@@ -32,15 +36,33 @@ namespace dawn {
 namespace iir {
 
 DoMethod::DoMethod(Interval interval, const StencilMetaInformation& metaData)
-    : interval_(interval), id_(IndexGenerator::Instance().getIndex()), metaData_(metaData) {}
+    : interval_(interval), id_(IndexGenerator::Instance().getIndex()), metaData_(metaData),
+      ASTStmtToSAPMap_() {}
+
+DoMethod::DoMethod(Interval interval, const StencilMetaInformation& metaData,
+                   const std::shared_ptr<AST> ast)
+    : interval_(interval), id_(IndexGenerator::Instance().getIndex()), metaData_(metaData),
+      ast_(ast), ASTStmtToSAPMap_() {}
+
+DoMethod::DoMethod(Interval interval, const StencilMetaInformation& metaData,
+                   const std::shared_ptr<AST> ast, const std::shared_ptr<SIR>& fullSIR,
+                   iir::StencilInstantiation* instantiation,
```

Comment:
I think is always better not to leave the shared ptr context and start passing raw pointers to functions.
Best would be `const std::shared_ptr<>& ` if not, just a `const &` . Last resort would be *, since they are very unsafe

Reply:
`iir::StencilInstantiation* instantiation` is required by the constructor of StatementMapper. Since I didn't want to change StatementMapper too much i kept it this way. 
However, now i realize adapting StatementMapper is easy, so I'll refactor it to use shared_ptr of instantiation_ instead of raw pointers. Still, we cannot use `const std::shared_ptr<>&` for stencil instantiation, as StatementMapper is going to call non-const methods of it.